pkg/controller/sparkapplication/submission.go

```diff
@@ -367,3 +380,70 @@ func addExecutorConfOptions(app *v1beta2.SparkApplication, submissionID string)
 
 	return executorConfOptions, nil
 }
+
+// addLocalDirConfOptions excludes local dir volumes, update SparkApplication and returns local dir config options
+func addLocalDirConfOptions(app *v1beta2.SparkApplication) ([]string, error) {
+	var localDirConfOptions []string
+
+	sparkLocalVolumes := map[string]v1.Volume{}
+	var mutateVolumes []v1.Volume
+
+	// Filter local dir volumes
+	for _, volume := range app.Spec.Volumes {
+		if strings.HasPrefix(volume.Name, config.SparkLocalDirVolumePrefix) {
+			sparkLocalVolumes[volume.Name] = volume
+		} else {
+			mutateVolumes = append(mutateVolumes, volume)
+		}
+	}
+	app.Spec.Volumes = mutateVolumes
+
+	// Filter local dir volumeMounts and set mutate volume mounts to driver and executor
+	if app.Spec.Driver.VolumeMounts != nil {
+		app.Spec.Driver.VolumeMounts = filterMutateMountVolumes(app.Spec.Driver.VolumeMounts, config.SparkDriverVolumesPrefix, sparkLocalVolumes, &localDirConfOptions)
+	}
+
+	if app.Spec.Executor.VolumeMounts != nil {
+		app.Spec.Executor.VolumeMounts = filterMutateMountVolumes(app.Spec.Executor.VolumeMounts, config.SparkExecutorVolumesPrefix, sparkLocalVolumes, &localDirConfOptions)
+	}
+
+	return localDirConfOptions, nil
+}
+
+func filterMutateMountVolumes(volumeMounts []v1.VolumeMount, prefix string, sparkLocalVolumes map[string]v1.Volume, localDirConfOptions *[]string) []v1.VolumeMount {
+	var mutateMountVolumes []v1.VolumeMount
+	for _, volumeMount := range volumeMounts {
+		if volume, ok := sparkLocalVolumes[volumeMount.Name]; ok {
+			options := buildLocalVolumeOptions(prefix, volume, volumeMount)
+			for _, option := range options {
+				*localDirConfOptions = append(*localDirConfOptions, option)
```

Comment:
Looks like you don't need to pass in `localDirConfOptions` as a pointer type to a slice, you can create it locally and return it instead.

Reply:
Cool. I make the update